bookstore/clone.py

```diff
@@ -299,3 +300,21 @@ def build_post_response_model(self, model, obj, s3_bucket, s3_object_key):
         if 'VersionId' in obj:
             model["versionID"] = obj['VersionId']
         return model
+
+
+def validate_relpath(relpath, settings):
+    """Validates that a relative path appropriately resolves given bookstore settings.
+    """
+    if relpath == '':
+        raise web.HTTPError(400, "Requires a path to clone")
+
+    fs_basedir = Path(settings.fs_cloning_basedir)
+    if not fs_basedir.is_absolute():
+        raise web.HTTPError(400, "File system cloning base directory must be an absolute path")
+
+    fs_clonepath = Path(os.path.realpath(os.path.join(fs_basedir, relpath)))
+
+    if fs_basedir not in fs_clonepath.parents:
+        raise web.HTTPError(400, "Cannot clone from a path outside of the base directory")
```

Comment:
Suggest making return the same http response as if the if the file did not exist, with only log messages that clarify why we rejected it

Reply:
We have different responses for a file not existing and a file being outside of the notebook basedir on the main notebook server. I'm going to model the response off of the response for a file being outside of the notebook basedir.

https://github.com/jupyter/notebook/blob/e498de6775b3de57f5ff827e562c179b17893064/notebook/services/contents/fileio.py#L262-L263
```python
        if not (os.path.abspath(os_path) + os.path.sep).startswith(root):
            raise HTTPError(404, "%s is outside root contents directory" % path)
```

Reply:
Ok, if that's something you are already doing then it's fine i guess.

Reply:
I still switched it to 404 and added the logging.